internal/integration/api/pcirebind.go

```diff
@@ -0,0 +1,173 @@
+		suite.T().Skip("skipping virtio test since provisioner is not qemu")
+	}
+
+	node := suite.RandomDiscoveredNodeInternalIP(machine.TypeWorker)
+
+	nodeCtx := client.WithNode(suite.ctx, node)
+
+	stream, err := suite.Client.LS(nodeCtx, &machineapi.ListRequest{
+		Root: "/sys/class/iommu",
+	})
+
+	suite.Require().NoError(err)
+
+	var count int
+
+	suite.Require().NoError(helpers.ReadGRPCStream(stream, func(info *machineapi.FileInfo, node string, multipleNodes bool) error {
+		if info.GetRelativeName() == "." {
+			return nil
+		}
+
+		count++
+
+		return nil
+	}))
+
+	if count == 0 {
+		suite.T().Skip("skipping PCI rebind test since IOMMU is not enabled")
+	}
+}
+
+// TearDownTest tears down the test.
+func (suite *PCIRebindSuite) TearDownTest() {
+	if suite.ctxCancel != nil {
+		suite.ctxCancel()
+	}
+}
+
+// TestIOMMURebind tests PCI rebind.
+func (suite *PCIRebindSuite) TestIOMMURebind() {
+	node := suite.RandomDiscoveredNodeInternalIP(machine.TypeWorker)
+
+	nodeCtx := client.WithNode(suite.ctx, node)
+
+	items, err := suite.Client.COSI.List(nodeCtx, resource.NewMetadata(hardware.NamespaceName, hardware.PCIDeviceType, "", resource.VersionUndefined))
+	suite.Require().NoError(err)
+
+	var pciDeviceID string
+
+	for _, item := range items.Items {
+		pci, ok := item.(*hardware.PCIDevice)
+		suite.Require().True(ok, "expected PCI device, got %T", item)
+
+		if pci.TypedSpec().Product == "Virtio 1.0 network device" {
+			pciDeviceID = pci.Metadata().ID()
+
+			break
+		}
+	}
+
+	suite.validateDriver(nodeCtx, pciDeviceID, "virtio-pci")
+
+	cfgDocument := pci.NewRebindConfigV1Alpha1()
+	cfgDocument.MetaName = pciDeviceID
+	cfgDocument.PCIVendorDeviceID = pciDeviceID
+	cfgDocument.PCITargetDriver = "vfio-pci"
+
+	v1alpha1CfgDocument := &v1alpha1.Config{
+		MachineConfig: &v1alpha1.MachineConfig{
+			MachineKernel: &v1alpha1.KernelConfig{
+				KernelModules: []*v1alpha1.KernelModuleConfig{
+					{
+						ModuleName: "vfio-pci",
+					},
+				},
+			},
+		},
+	}
+
+	suite.PatchMachineConfig(nodeCtx, v1alpha1CfgDocument, cfgDocument)
+
+	_, err = suite.Client.COSI.WatchFor(nodeCtx, runtimeres.NewPCIRebindStatus(pciDeviceID).Metadata(), state.WithEventTypes(state.Created, state.Updated))
+	suite.Require().NoError(err)
+
+	suite.validateDriver(nodeCtx, pciDeviceID, "vfio-pci")
+
+	cfgDocument.PCITargetDriver = "virtio-pci"
+
+	suite.PatchMachineConfig(nodeCtx, cfgDocument)
+
+	_, err = suite.Client.COSI.WatchFor(nodeCtx, runtimeres.NewPCIRebindStatus(pciDeviceID).Metadata(), state.WithEventTypes(state.Updated))
+	suite.Require().NoError(err)
+
+	suite.validateDriver(nodeCtx, pciDeviceID, "virtio-pci")
+
+	suite.RemoveMachineConfigDocuments(nodeCtx, cfgDocument.MetaKind)
+
+	_, err = suite.Client.COSI.WatchFor(nodeCtx, runtimeres.NewPCIRebindStatus(pciDeviceID).Metadata(), state.WithEventTypes(state.Destroyed))
+	suite.Require().NoError(err)
+
+	suite.validateDriver(nodeCtx, pciDeviceID, "virtio-pci")
```

Comment:
Maybe add transition to `vfio-pci` once more before this, so we're sure it also rebinds back to the original on config document removal? As this doesn't really test this since you first ask it to switch to the original driver by config

Reply:
yup, good point